src/release_notes_generator.py

```diff
@@ -0,0 +1,196 @@
+    :return: The generated release notes as a string, or None if the repository could not be found.
+    """
+    # get GitHub repository object (1 API call)
+    if GithubManager().fetch_repository(repository_id) is None:
+        return None
+
+    # get latest release (1 API call)
+    GithubManager().fetch_latest_release()
+    GithubManager().show_rate_limit()
+
+    # get closed issues since last release (N API calls - pagination)
+    issues = GithubManager().fetch_issues()
+    GithubManager().show_rate_limit()
+
+    # get finished PRs since last release
+    pulls = GithubManager().fetch_pull_requests(state='closed')
+    GithubManager().show_rate_limit()
+
+    # get commits since last release
+    commits = GithubManager().fetch_commits()
+
+    # generate change url
+    changelog_url = GithubManager().get_change_url(tag_name)
+
+    # merge data to Release Notes records form
+    rls_notes_records: dict[int, Record] = RecordFactory.generate(
+        issues=issues,
+        pulls=pulls,
+        commits=commits
+    )
+
+    formatter = RecordFormatter()
+
+    # build rls notes
+    return ReleaseNotesBuilder(
+        records=rls_notes_records,
+        custom_chapters=custom_chapters,
+        formatter=formatter,
+        warnings=warnings,
+        print_empty_chapters=print_empty_chapters,
+        changelog_url=changelog_url
+    ).build()
+
+
+def run():
+    """
+    Runs the 'Release Notes Generator' GitHub Action.
+    """
+    logging.info("Starting 'Release Notes Generator' GitHub Action")
+
+    try:
+        local_repository_id = get_action_input('GITHUB_REPOSITORY')
+        owner, repo_name = local_repository_id.split('/')
+
+        github_token: str = get_action_input('github-token')
+
+        tag_name = get_action_input('tag-name')
+        chapters_json = get_action_input('chapters')
+        warnings = get_action_input('warnings') == 'true'
+        published_at = get_action_input('published-at') == 'true'
+        skip_release_notes_label_raw = get_action_input('skip-release-notes-label')
+        skip_release_notes_label = skip_release_notes_label_raw if skip_release_notes_label_raw else 'skip-release-notes'
+        print_empty_chapters = get_action_input('print-empty-chapters') == 'true'
+        chapters_to_pr_without_issue = get_action_input('chapters-to-pr-without-issue') == 'true'
+        verbose = get_action_input('verbose').lower() == 'true'
+        if verbose:
+            logging.info("Verbose logging enabled")
+            logging.getLogger().setLevel(logging.DEBUG)
+        else:
+            logging.info("Verbose logging disabled")
+
+        # Init GitHub instance
+        auth = Auth.Token(token=github_token)
+        g = Github(auth=auth, per_page=100)
+        GithubManager().github = g    # creat singleton instance and init with g (Github)
+        GithubManager().show_rate_limit()
+
+        validate_inputs(owner, repo_name, tag_name, chapters_json, warnings, published_at,
```

Comment:
also the validation could be moved there into such class, not here

Reply:
Thx. Addressed in commit 510b0c7e42e2dca52459d023a16c7771eeed784e.